plutus-tx/compiler/Language/PlutusTx/Compiler/Expr.hs

```diff
@@ -314,7 +316,21 @@ compileExpr e = withContextM 2 (sdToTxt $ "Compiling expr:" GHC.<+> GHC.ppr e) $
         GHC.Var (GHC.idDetails -> GHC.PrimOpId po) -> compilePrimitiveOp po
         GHC.Var (GHC.idDetails -> GHC.DataConWorkId dc) -> compileDataConRef dc
         -- See Note [Unfoldings]
-        GHC.Var n@(GHC.realIdUnfolding -> GHC.CoreUnfolding{GHC.uf_tmpl=unfolding}) -> hoistExpr n unfolding
+        -- The "unfolding template" includes things with normal unfoldings and also dictionary functions
+        GHC.Var n@(GHC.maybeUnfoldingTemplate . GHC.realIdUnfolding -> Just unfolding) -> hoistExpr n unfolding
+        -- Class ops don't have unfoldings in general (although they do if they're for one-method classes, so we
+        -- want to check the unfoldings case first), see the GHC Note [ClassOp/DFun selection] for why. That
+        -- means we have to reconstruct the RHS ourselves, though, which is a pain.
+        GHC.Var n@(GHC.idDetails -> GHC.ClassOpId cls) -> do
+            -- This code (mostly) lifted from MkId.mkDictSelId, which makes unfoldings for those dictionary
+            -- selectors that do have them
+            let sel_names = fmap GHC.getName (GHC.classAllSelIds cls)
+            val_index <- case elemIndex (GHC.getName n) sel_names of
+                Just i  -> pure i
+                Nothing -> throwPlain $ CompilationError "Id not in class method list"
```

Comment:
Can `n` be included in the error?

Reply:
Yep, this is "impossible", but always good to have useful errors when the impossible inevitably happens.